cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/RemoteProgramRunDispatcher.java

```diff
@@ -118,13 +115,18 @@ public ProgramController dispatchProgram(ProgramRunDispatcherInfo programRunDisp
         programId.getType());
     RunRecordDetail runRecordDetail = store.getRun(programRunId);
     TwillController twillController = null;
-    if (remoteTwillRunnerService != null && runRecordDetail != null) {
+    if (runRecordDetail != null) {
       twillController = remoteTwillRunnerService.createTwillControllerFromRunRecord(runRecordDetail);
     }
     ProgramController programController = null;
-    if (runner instanceof DistributedWorkflowProgramRunner && twillController != null) {
-      programController =
-        ((DistributedWorkflowProgramRunner) runner).createProgramController(programRunId, twillController);
+    if (runner instanceof ProgramControllerCreator && twillController != null) {
+      programController = ((ProgramControllerCreator) runner).createProgramController(programRunId, twillController);
+    }
+    if (Objects.isNull(programController)) {
+      String msg = String.format("Unable to dispatch Program %s with runid %s",
```

Comment:
this message needs more detail for it to be useful. For example, if the run record doesn't exist, running it again can succeed. But if the runner isn't a ProgramControllerCreator for some reason, that will always fail and indicates it's some code bug. It would be better to do something like:

```
if (runRecordDetail == null) {
  throw new IllegalStateException("Could not find run record for ...");
}
...
if (!(runner instanceof ProgramControllerCreator)) {
  throw new IllegalStateException("Program ... uses an unsupported controller for remote dispatching.");
}
...
```